contracts/mixnet/src/helpers.rs

```diff
@@ -65,11 +73,181 @@ pub(crate) fn scale_reward_by_uptime(
     Ok(uint128_to_decimal(scaled))
 }
 
+// Extracts the node identity and owner of a delegation from the bytes used as
+// key in the delegation buckets.
+fn extract_identity_and_owner(bytes: Vec<u8>) -> StdResult<(Addr, IdentityKey)> {
+    if bytes.len() < NAMESPACE_LENGTH {
+        return Err(StdError::parse_err(
+            "mixnet_contract::types::IdentityKey",
+            "Invalid type",
+        ));
+    }
+    let identity_size = u16::from_be_bytes([bytes[0], bytes[1]]) as usize;
+    let identity_bytes: Vec<u8> = bytes
+        .iter()
+        .skip(NAMESPACE_LENGTH)
+        .take(identity_size)
+        .copied()
+        .collect();
+    let identity = IdentityKey::from_utf8(identity_bytes)
+        .map_err(|_| StdError::parse_err("mixnet_contract::types::IdentityKey", "Invalid type"))?;
+    let owner_bytes: Vec<u8> = bytes
+        .iter()
+        .skip(NAMESPACE_LENGTH + identity_size)
+        .copied()
+        .collect();
+    let owner = Addr::unchecked(
+        String::from_utf8(owner_bytes)
+            .map_err(|_| StdError::parse_err("cosmwasm_std::addresses::Addr", "Invalid type"))?,
+    );
+
+    Ok((owner, identity))
+}
+
+// currently not used outside tests
+#[cfg(test)]
+// Converts the node identity and owner of a delegation into the bytes used as
+// key in the delegation buckets.
+pub(crate) fn identity_and_owner_to_bytes(identity: &IdentityKey, owner: &Addr) -> Vec<u8> {
+    let mut bytes = u16::to_be_bytes(identity.len() as u16).to_vec();
+    bytes.append(&mut identity.as_bytes().to_vec());
+    bytes.append(&mut owner.as_bytes().to_vec());
+
+    bytes
+}
+
+pub(crate) fn get_all_delegations_paged<T>(
+    bucket: &ReadonlyBucket<T>,
+    start_after: &Option<Vec<u8>>,
+    limit: usize,
+) -> StdResult<PagedAllDelegationsResponse<T>>
+where
+    T: Serialize + DeserializeOwned,
+{
+    let delegations = bucket
+        .range(start_after.as_deref(), None, Order::Ascending)
+        .filter(|res| res.is_ok())
+        .take(limit)
+        .map(|res| {
+            res.map(|entry| {
+                let (owner, identity) = extract_identity_and_owner(entry.0).expect("Invalid node identity or address used as key in bucket. The storage is corrupted!");
+                UnpackedDelegation::new(owner, identity, entry.1)
+            })
+        })
+        .collect::<StdResult<Vec<UnpackedDelegation<T>>>>()?;
+
+    let start_next_after = bucket
+        .range(start_after.as_deref(), None, Order::Ascending)
+        .filter(|res| res.is_ok())
+        .take(limit)
+        .last()
+        .map(|res| res.unwrap().0);
```

Comment:
Is there a scenario in which `filter` returns an empty iterator? I'm a bit worried about that `unwrap`, could we do an `if let` instead of the `map`?

Reply:
If `filter` returns an empty iterator, `take` will also return an empty iterator and `last` will return `None`, so `map`'s lambda won't even get invoked. The `unwrap` from the `map` is safe because of the above `filter` that checks all elements can safely be unwrapped.
That being said, I will change this approach to a more readable one, using `if let`.